### PR TITLE
Added support for interpolated variables

### PR DESCRIPTION
**Pull request by bipolar**
It adds support for interpolated variables like 
var1 = value1
var2 = ${var1}-value2
It will also expand ${sys:...} and ${env:...} keys.

As a draw back it increase the size of the pull-in because of the extra three libraries form Apache Commons.
Reference : http://commons.apache.org/proper/commons-configuration/userguide/howto_utilities.html#Utility_classes_and_Tips_and_Tricks

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -68,6 +68,16 @@
...
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-configuration2</artifactId>
+            <version>2.2</version>
+        </dependency>
+        <dependency>
+            <groupId>commons-beanutils</groupId>
+            <artifactId>commons-beanutils</artifactId>
+            <version>1.9.3</version>
+        </dependency>
```
Comment (rsandell) [inline pom.xml]: @jglick @oleg-nenashev do you have any insight if these dependencies are safe to add?
´commons-beanutils´ is provided by code IIRC?

Comment (jglick) [reply pom.xml]: Cannot recall offhand, but should be checked.

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/ReadPropertiesStepExecution.java [hunk 2]
```diff
@@ -82,6 +87,9 @@
             properties.load(sr);
         }
 
+        // Interpolated values in the properties
+        properties = interpolateProperties(properties);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/ReadPropertiesStepExecution.java]: I do not recommend changing the behavior of the `readProperties` step. Rather introduce a new step for interpolation if you want one. Or just leave this to a global library with `@Grab` for people who need it.

Comment (rsandell) [reply src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/ReadPropertiesStepExecution.java]: Or maybe adding a new parameter telling the step to change behavior?

Comment (bipolar) [reply src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/ReadPropertiesStepExecution.java]: Sounds better with a parameter, the impact on the code is minimal.

File: src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/conf/ReadPropertiesStep/help.html [hunk 3]
```diff
@@ -45,6 +45,11 @@
         An Optional Map containing default key/values.
         <em>These are added to the resulting map first.</em>
     </li>
+    <li>
+        <code>interpolate</code>:
+        Flag to indicate if the properties should be interpolated or not.
+        In case of error or cycling dependencies, the original properties will be returned.
```
Comment (rsandell) [inline src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/conf/ReadPropertiesStep/help.html]: An example would be nice here. Not all users understands what _interpolated_ means :)

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/ReadPropertiesStep.java [hunk 4]
```diff
@@ -39,6 +39,7 @@
  */
 public class ReadPropertiesStep extends AbstractFileOrTextStep {
     private Map defaults;
+    private Boolean interpolate = Boolean.FALSE;
```
Comment (rsandell) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/ReadPropertiesStep.java]: a normal, basic `boolean` should suffice here, it defaults to `false` in Java.

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/ReadPropertiesStep.java [hunk 5]
```diff
@@ -63,6 +64,27 @@ public void setDefaults(Map defaults) {
...
 
+    /**
+     * Flag to indicate if the properties should be interpolated or not.
+     * I.E. :
+     *   baseUrl = http://localhost
+     *   url = ${baseUrl}/resources
+     * The value of <i>url</i> should be evaluated to http://localhost/resources with the interpolation on.
+     * @return the value of interpolated
+     */
+    public Boolean getInterpolate() {
```
Comment (rsandell) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/ReadPropertiesStep.java]: following the Java bean spec this should then be `isInterpolate`.

File: pom.xml (hunk 1 above)
Comment (rsandell) [reply pom.xml]: @bipolar please check

Comment (bipolar) [reply pom.xml]: Indeed jenkins-core already provides beanutils, I've removed the dependency.

File: src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/conf/ReadPropertiesStep/help.html (hunk 3 above)
Comment (bipolar) [reply src/main/resources/org/jenkinsci/plugins/pipeline/utility/steps/conf/ReadPropertiesStep/help.html]: Done! I've added an extra example with interpolation.

File: src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/ReadPropertiesStepExecution.java [hunk 6]
```diff
@@ -103,4 +112,30 @@ private void addAll(Map src, Map<String, Object> dst) {
...
+    /**
+     * Using commons collection to interpolated the values inside the properties
+     * @param properties the list of properties to be interpolated
+     * @return a new Properties object with the interpolated values
+     */
+    private Properties interpolateProperties(Properties properties) {
+        if ( properties == null)
+            return null;
+        Configuration interpolatedProp;
+        PrintStream logger = listener.getLogger();
```
Comment (rsandell) [inline src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/ReadPropertiesStepExecution.java]: You need to rebase your master branch. The `listener` field has been removed so you are getting a compilation failure here.

**Pull request merged**